File: src/game/training/trainingQuestionBuilder.ts
import { distance, transpose } from 'tonal';

import type {
  TrainingQuestion,
  TrainingQuestionBuilderOptions,
  TrainingQuestionNote,
} from '@/game/training/trainingTypes';
import { CHORD_TEMPLATES, SCALE_TEMPLATES } from '@/utils/chord-templates';
import type { ChordQuality, ScaleType } from '@/utils/chord-templates';
import {
  getNotationInstrumentPreset,
  getWrittenSemitoneOffset,
  normalizeNotationInstrumentId,
} from '@/utils/notationInstrument';
import { parseVoicingNoteName } from '@/utils/voicingMusicXml';

type Clef = 'treble' | 'bass';
type StaffNumber = 1 | 2;

const NATURAL_PITCH_CLASSES: ReadonlySet<number> = new Set([0, 2, 4, 5, 7, 9, 11]);

const naturalMidisInRange = (min: number, max: number): number[] => {
  const out: number[] = [];
  for (let midi = min; midi <= max; midi += 1) {
    if (NATURAL_PITCH_CLASSES.has(((midi % 12) + 12) % 12)) out.push(midi);
  }
  return out;
};

/** 譜読み: 加線1本までの幹音（記譜音）。ト音 C4-A5 / ヘ音 E2-C4 */
const NOTE_READING_CANDIDATES: Record<Clef, readonly number[]> = {
  treble: naturalMidisInRange(60, 81),
  bass: naturalMidisInRange(40, 60),
};

/** コンサート音高を♭系の綴りで音名にする（移調楽器は全て♭系のため記譜上は幹音になる） */
const FLAT_PITCH_NAMES = ['C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab', 'A', 'Bb', 'B'] as const;
const flatSpelledName = (midi: number): string => {
  const pc = ((midi % 12) + 12) % 12;
  return `${FLAT_PITCH_NAMES[pc] ?? 'C'}${Math.floor(midi / 12) - 1}`;
};

/** 和音・スケール・音程: 5線内に収まる最低音（記譜音）。ト音 E4 / ヘ音 G2 */
const STAFF_BOTTOM_MIDI: Record<Clef, number> = {
  treble: 64,
  bass: 43,
};

/** 音程トレーニングの基準音候補（綴りを固定） */
const INTERVAL_BASE_SPELLINGS: readonly string[] = [
  'C', 'C#', 'Db', 'D', 'D#', 'Eb', 'E', 'F', 'F#', 'Gb', 'G', 'G#', 'Ab', 'A', 'A#', 'Bb', 'B',
];

/** 出題コードネーム用のクオリティ表記 */
const CHORD_SYMBOL_SUFFIX: Partial<Record<ChordQuality, string>> = {
  maj: '',
  min: 'm',
  dim: 'dim',
  aug: 'aug',
  sus4: 'sus4',
  maj7: 'M7',
  m7: 'm7',
  '7': '7',
  m7b5: 'm7(b5)',
  dim7: 'dim7',
  '7sus4': '7sus4',
  '6': '6',
  m6: 'm6',
  mM7: 'mM7',
};

const pickRandom = <T,>(items: readonly T[]): T => {
  if (items.length === 0) {
    throw new Error('pickRandom: empty');
  }
  return items[Math.floor(Math.random() * items.length)] as T;
};

const normalizePitchClass = (midi: number): number => ((midi % 12) + 12) % 12;

const accidentalText = (alter: number): string => {
  if (alter === 2) return 'x';
  if (alter === 1) return '#';
  if (alter === -1) return 'b';
  if (alter === -2) return 'bb';
  return '';
};

const normalizeSpelling = (name: string): string => name.replace(/##/g, 'x');

const midiOf = (name: string): number => parseVoicingNoteName(name).midi;

/** 綴りを保ってオクターブだけずらす */
const shiftOctave = (name: string, delta: number): string => {
  const parsed = parseVoicingNoteName(name);
  return `${parsed.step}${accidentalText(parsed.alter)}${parsed.octave + delta}`;
};

/** 最低音が [minMidi, minMidi + 12) に入る最も低いオクターブへ全音を平行移動する */
const placeLowestInOctaveAbove = (names: readonly string[], minMidi: number): string[] => {
  if (names.length === 0) return [];
  let lowest = Number.POSITIVE_INFINITY;
  for (const name of names) {
    const midi = midiOf(name);
    if (midi < lowest) lowest = midi;
  }
  const octaveDelta = Math.ceil((minMidi - lowest) / 12);
  if (octaveDelta === 0) return names.slice();
  return names.map((name) => shiftOctave(name, octaveDelta));
};

const spelledFromIntervals = (rootWithOctave: string, intervals: readonly string[]): string[] =>
  intervals.map((interval) => {
    const note = transpose(rootWithOctave, interval);
    if (!note) {
      throw new Error(`transpose failed: ${rootWithOctave} + ${interval}`);
    }
    return normalizeSpelling(note);
  });

const transposeVoicingToRoot = (
  voicingNotes: readonly string[],
  referenceRoot: string,
  targetRoot: string,
): string[] => {
  const interval = distance(referenceRoot, targetRoot);
  if (!interval || interval === '1P') {
    return voicingNotes.slice();
  }
  return voicingNotes.map((name) => {
    const transposed = transpose(name, interval);
    if (!transposed) {
      throw new Error(`transpose failed: ${name} + ${interval}`);
    }
    return normalizeSpelling(transposed);
  });
};

/** 最低音の直下にあるルート音（正解時に鳴らす音） */
const rootMidiBelow = (root: string, lowestMidi: number): number => {
  const rootPc = normalizePitchClass(midiOf(`${root}4`));
  let midi = lowestMidi - ((normalizePitchClass(lowestMidi) - rootPc + 12) % 12);
  if (midi >= lowestMidi) midi -= 12;
  return midi;
};

/** 音程トレーニングでは重変化記号と E# / B# / Cb / Fb を避ける */
const isSimpleSpelling = (name: string): boolean => {
  const parsed = parseVoicingNoteName(name);
  if (Math.abs(parsed.alter) > 1) return false;
  if (parsed.alter === 1 && (parsed.step === 'E' || parsed.step === 'B')) return false;
  if (parsed.alter === -1 && (parsed.step === 'C' || parsed.step === 'F')) return false;
  return true;
};

const resolveEffectiveClef = (
  trainingClefMode: TrainingQuestionBuilderOptions['training']['clefMode'],
  instrumentClef: 'treble' | 'bass' | 'grand',
  configClef: 'auto' | 'treble' | 'bass' | undefined,
): Clef | 'grand' => {
  if (trainingClefMode === 'bass_concert') return 'bass';
  if (trainingClefMode === 'grand_concert') return 'grand';
  if (configClef === 'treble') return 'treble';
  if (configClef === 'bass') return 'bass';
  return instrumentClef === 'grand' ? 'treble' : instrumentClef;
};

const buildNotes = (
  names: readonly string[],
  staves: readonly StaffNumber[],
  targets: readonly boolean[],
): TrainingQuestionNote[] => names.map((noteName, index) => {
  const midi = midiOf(noteName);
  return {
    noteName,
    midi,
    pitchClass: normalizePitchClass(midi),
    staff: staves[index] ?? 1,
    isTarget: targets[index] ?? true,
  };
});

const toStaves = (staves: readonly number[] | undefined, count: number, fallback: StaffNumber): StaffNumber[] => {
  const out: StaffNumber[] = [];
  for (let i = 0; i < count; i += 1) {
    const s = staves?.[i];
    out.push(s === 2 ? 2 : s === 1 ? 1 : fallback);
  }
  return out;
};

const allTrue = (count: number): boolean[] => Array.from({ length: count }, () => true);

export const buildTrainingQuestion = (
  options: TrainingQuestionBuilderOptions,
): TrainingQuestion => {
  const { training, previousQuestionKey } = options;
  const preset = getNotationInstrumentPreset(normalizeNotationInstrumentId(options.notationInstrumentId));
  const writtenOffset = options.ignoreNotationInstrument
    ? 0
    : getWrittenSemitoneOffset(preset, options.notationOctaveShift);
  // 調号は常にコンサート C（0）。移調楽器の調号は ChordVoicingStaff 側で writtenOffset から算出する。
  const keyFifths = 0;
  const config = options.lessonItems && options.lessonItems.length > 0
    ? { ...training.config, ...options.lessonItems[options.lessonItemIndex ?? 0] }
    : training.config;

  const roots = options.lessonRoots ?? config.roots ?? ['C'];
  const root = options.lessonOrder === 'sequential' && options.lessonItemIndex != null
    ? roots[options.lessonItemIndex % roots.length] ?? 'C'
    : pickRandom(roots);

  const effectiveClef = resolveEffectiveClef(training.clefMode, preset.clef, config.clef);
  const singleClef: Clef = effectiveClef === 'bass' ? 'bass' : 'treble';
  const defaultStaff: StaffNumber = singleClef === 'bass' ? 2 : 1;
  // 記譜上の5線最低音をコンサート音高に戻す
  const concertStaffBottom = STAFF_BOTTOM_MIDI[singleClef] - writtenOffset;

  const makeQuestion = (
    questionKey: string,
    promptLabel: string,
    noteNames: readonly string[],
    staves: readonly StaffNumber[],
    targets: readonly boolean[],
    layout: 'stacked' | 'horizontal',
    ordered: boolean,
    rootMidi: number | null,
  ): TrainingQuestion => ({
    questionKey,
    promptLabel,
    notes: buildNotes(noteNames, staves, targets),
    layout,
    ordered,
    keyFifths,
    rootMidi,
  });

  for (let attempt = 0; attempt < 12; attempt += 1) {
    if (training.kind === 'note_reading') {
      // 記譜上の白鍵（幹音）のみを出題する
      const writtenMidi = pickRandom(NOTE_READING_CANDIDATES[singleClef]);
      const concertMidi = writtenMidi - writtenOffset;
      const questionKey = `note:${concertMidi}`;
      if (questionKey === previousQuestionKey) continue;
      return makeQuestion(
        questionKey,
        '',
        [flatSpelledName(concertMidi)],
        [defaultStaff],
        [true],
        'stacked',
        false,
        concertMidi,
      );
    }

    if (training.kind === 'interval') {
      const interval = config.interval ?? '2m';
      const direction = config.direction ?? 'up';
      const tonalInterval = direction === 'up' ? interval : `-${interval}`;
      const candidates: { base: string; target: string }[] = [];
      for (const spelling of INTERVAL_BASE_SPELLINGS) {
        const [base] = placeLowestInOctaveAbove([`${spelling}4`], concertStaffBottom);
        if (!base) continue;
        const target = transpose(base, tonalInterval);
        if (!target) continue;
        const normalizedTarget = normalizeSpelling(target);
        if (!isSimpleSpelling(normalizedTarget)) continue;
        if (midiOf(normalizedTarget) < concertStaffBottom) {
          candidates.push({ base: shiftOctave(base, 1), target: shiftOctave(normalizedTarget, 1) });
        } else {
          candidates.push({ base, target: normalizedTarget });
        }
      }
      if (candidates.length === 0) {
        throw new Error(`Training ${training.slug}: no interval candidates`);
      }
      const picked = pickRandom(candidates);
      const questionKey = `interval:${picked.base}:${picked.target}`;
      if (questionKey === previousQuestionKey) continue;
      return makeQuestion(
        questionKey,
        training.titleJa,
        [picked.base, picked.target],
        [defaultStaff, defaultStaff],
        [false, true],
        'stacked',
        false,
        midiOf(picked.base),
      );
    }

    if (training.kind === 'scale') {
      const scaleType = (config.scale ?? 'major') as ScaleType;
      const intervals = SCALE_TEMPLATES[scaleType];
      if (!intervals) {
        throw new Error(`Training ${training.slug}: unknown scale ${scaleType}`);
      }
      const names = placeLowestInOctaveAbove(spelledFromIntervals(`${root}4`, intervals), concertStaffBottom);
      const questionKey = `scale:${root}:${scaleType}`;
      if (questionKey === previousQuestionKey) continue;
      return makeQuestion(
        questionKey,
        `${root} ${training.titleJa}`,
        names,
        toStaves(undefined, names.length, defaultStaff),
        allTrue(names.length),
        'horizontal',
        true,
        midiOf(names[0] ?? `${root}4`),
      );
    }

    if (training.kind === 'chord' || training.kind === 'voicing') {
      let names: string[];
      let staves: StaffNumber[];

      if (config.voicingNotes && config.voicingNotes.length > 0) {
        const transposed = transposeVoicingToRoot(config.voicingNotes, config.referenceRoot ?? 'C', root);
        names = config.minLowestNote
          ? placeLowestInOctaveAbove(transposed, midiOf(config.minLowestNote))
          : placeLowestInOctaveAbove(transposed, concertStaffBottom);
        staves = toStaves(config.staves, names.length, defaultStaff);
      } else if (config.intervals && config.intervals.length > 0) {
        const raw = spelledFromIntervals(`${root}3`, config.intervals);
        names = placeLowestInOctaveAbove(
          raw,
          config.minLowestNote ? midiOf(config.minLowestNote) : concertStaffBottom,
        );
        staves = toStaves(config.staves, names.length, defaultStaff);
      } else if (config.quality) {
        const intervals = CHORD_TEMPLATES[config.quality];
        if (!intervals) {
          throw new Error(`Training ${training.slug}: unknown quality ${config.quality}`);
        }
        names = placeLowestInOctaveAbove(spelledFromIntervals(`${root}4`, intervals), concertStaffBottom);
        staves = toStaves(undefined, names.length, defaultStaff);
      } else {
        throw new Error(`Training ${training.slug}: missing chord/voicing config`);
      }

      const questionKey = `chord:${root}:${names.join('|')}`;
      if (questionKey === previousQuestionKey) continue;
      const suffix = training.kind === 'chord' && config.quality
        ? CHORD_SYMBOL_SUFFIX[config.quality] ?? config.quality
        : training.titleEn;
      const lowestMidi = Math.min(...names.map(midiOf));
      return makeQuestion(
        questionKey,
        `${root}${suffix}`,
        names,
        staves,
        allTrue(names.length),
        'stacked',
        false,
        rootMidiBelow(root, lowestMidi),
      );
    }

    throw new Error(`Unsupported training kind: ${training.kind}`);
  }

  return buildTrainingQuestion({ ...options, previousQuestionKey: null });
};

export const createInitialTrainingRuntime = (): import('@/game/training/trainingTypes').TrainingRuntime => ({
  durationSec: 60,
  elapsedSec: 0,
  score: 0,
  result: 'playing',
  enemy: { typeIndex: 0, active: true, fadeAlpha: 1, slashUntilSec: 0 },
  question: null,
  correctTargetIndices: [],
  nextQuestionKey: null,
});

File: ios/Jazzify/Training/Engine/TrainingQuestionBuilder.swift
import Foundation

/// Web `trainingQuestionBuilder.ts` と同じ出題ロジック。
/// iOS は記譜楽器プリセットが無いため常にコンサート表記（writtenOffset = 0）。
enum TrainingQuestionBuilder {
    private typealias SpelledNote = TrainingMusicTheory.SpelledNote

    /// 和音・スケール・音程: 5線内に収まる最低音。ト音 E4 / ヘ音 G2
    private static let staffBottomMidi: [String: Int] = ["treble": 64, "bass": 43]

    private static let naturalPitchClasses: Set<Int> = [0, 2, 4, 5, 7, 9, 11]

    private static let noteReadingRanges: [String: ClosedRange<Int>] = [
        "treble": 60...81,
        "bass": 40...60,
    ]

    private static let spellingsByPc: [[String]] = [
        ["C"],
        ["C#", "Db"],
        ["D"],
        ["D#", "Eb"],
        ["E"],
        ["F"],
        ["F#", "Gb"],
        ["G"],
        ["G#", "Ab"],
        ["A"],
        ["A#", "Bb"],
        ["B"],
    ]

    private static func buildNoteReadingSpellings(clef: String, includeAccidentals: Bool) -> [String] {
        guard let range = noteReadingRanges[clef] else { return [] }
        var out: [String] = []
        for midi in range {
            let pc = ((midi % 12) + 12) % 12
            if !includeAccidentals && !naturalPitchClasses.contains(pc) { continue }
            let octave = midi / 12 - 1
            for spelling in spellingsByPc[pc] {
                out.append("\(spelling)\(octave)")
            }
        }
        return out
    }

    private static let noteReadingNaturals: [String: [String]] = [
        "treble": buildNoteReadingSpellings(clef: "treble", includeAccidentals: false),
        "bass": buildNoteReadingSpellings(clef: "bass", includeAccidentals: false),
    ]

    private static let noteReadingWithAccidentals: [String: [String]] = [
        "treble": buildNoteReadingSpellings(clef: "treble", includeAccidentals: true),
        "bass": buildNoteReadingSpellings(clef: "bass", includeAccidentals: true),
    ]

    /// 音程トレーニングの基準音候補（綴りを固定）
    private static let intervalBaseSpellings = [
        "C", "C#", "Db", "D", "D#", "Eb", "E", "F", "F#", "Gb", "G", "G#", "Ab", "A", "A#", "Bb", "B",
    ]

    static func buildQuestion(options: TrainingQuestionBuilderOptions) -> TrainingQuestion {
        let training = options.training
        let previousQuestionKey = options.previousQuestionKey
        // 調号は常にコンサート C（0）。
        let keyFifths = 0

        let mergedConfig: TrainingConfig
        if let lessonItems = options.lessonItems, !lessonItems.isEmpty {
            let index = options.lessonItemIndex ?? 0
            let lessonItem = lessonItems[index % lessonItems.count]
            mergedConfig = mergeConfig(base: training.config, override: lessonItem)
        } else {
            mergedConfig = training.config
        }

        let roots = options.lessonRoots ?? mergedConfig.roots ?? ["C"]
        let root: String
        if options.lessonOrder == "sequential", let index = options.lessonItemIndex {
            root = roots[index % roots.count]
        } else {
            root = roots.randomElement() ?? "C"
        }

        let effectiveClef = resolveEffectiveClef(clefMode: training.clefMode, configClef: mergedConfig.clef)
        let singleClef = effectiveClef == "bass" ? "bass" : "treble"
        let defaultStaff = singleClef == "bass" ? 2 : 1
        let staffBottom = staffBottomMidi[singleClef] ?? 64

        for _ in 0..<12 {
            switch training.kind {
            case .noteReading:
                let includeAccidentals = mergedConfig.includeAccidentals == true
                let forcedClef = mergedConfig.clef == "bass" || effectiveClef == "bass" ? "bass" : "treble"
                let spellings = includeAccidentals
                    ? noteReadingWithAccidentals[forcedClef]
                    : noteReadingNaturals[forcedClef]
                guard let writtenSpelling = spellings?.randomElement(),
                      let writtenMidi = TrainingMusicTheory.parseVoicingMidi(writtenSpelling)
                else { continue }
                let questionKey = "note:\(writtenMidi):\(writtenSpelling)"
                if questionKey == previousQuestionKey { continue }
                return makeQuestion(
                    questionKey: questionKey,
                    promptLabel: "",
                    noteNames: [writtenSpelling],
                    staves: [forcedClef == "bass" ? 2 : 1],
                    targets: [true],
                    layout: .stacked,
                    ordered: false,
                    keyFifths: keyFifths,
                    rootMidi: writtenMidi
                )

            case .interval:
                let interval = mergedConfig.interval ?? "2m"
                let directionUp = (mergedConfig.direction ?? "up") == "up"
                guard let spec = TrainingMusicTheory.parseInterval(directionUp ? interval : "-\(interval)") else { continue }
                var candidates: [(base: SpelledNote, target: SpelledNote)] = []
                for spelling in intervalBaseSpellings {
                    guard let raw = TrainingMusicTheory.parseSpelled("\(spelling)4") else { continue }
                    guard let base = TrainingMusicTheory.placeLowestInOctaveAbove([raw], minMidi: staffBottom).first else { continue }
                    let target = TrainingMusicTheory.transpose(base, by: spec)
                    guard TrainingMusicTheory.isSimpleSpelling(target) else { continue }
                    if target.midi < staffBottom {
                        candidates.append((
                            TrainingMusicTheory.shiftOctave(base, by: 1),
                            TrainingMusicTheory.shiftOctave(target, by: 1)
                        ))
                    } else {
                        candidates.append((base, target))
                    }
                }
                guard let picked = candidates.randomElement() else { continue }
                let questionKey = "interval:\(picked.base.name):\(picked.target.name)"
                if questionKey == previousQuestionKey { continue }
                return makeQuestion(
                    questionKey: questionKey,
                    promptLabel: training.titleJa,
                    noteNames: [picked.base.name, picked.target.name],
                    staves: [defaultStaff, defaultStaff],
                    targets: [false, true],
                    layout: .stacked,
                    ordered: false,
                    keyFifths: keyFifths,
                    rootMidi: picked.base.midi
                )

            case .scale:
                let scaleType = mergedConfig.scale ?? "major"
                guard let intervals = TrainingMusicTheory.scaleTemplates[scaleType] else { continue }
                let notes = TrainingMusicTheory.placeLowestInOctaveAbove(
                    TrainingMusicTheory.spelledFromIntervals(root: root, octave: 4, intervals: intervals),
                    minMidi: staffBottom
                )
                let questionKey = "scale:\(root):\(scaleType)"
                if questionKey == previousQuestionKey { continue }
                return makeQuestion(
                    questionKey: questionKey,
                    promptLabel: "\(root) \(training.titleJa)",
                    noteNames: notes.map(\.name),
                    staves: notes.map { _ in defaultStaff },
                    targets: notes.map { _ in true },
                    layout: .horizontal,
                    ordered: true,
                    keyFifths: keyFifths,
                    rootMidi: notes.first?.midi
                )

            case .chord, .voicing:
                guard let built = buildChordVoicingQuestion(
                    training: training,
                    config: mergedConfig,
                    root: root,
                    defaultStaff: defaultStaff,
                    staffBottom: staffBottom,
                    previousQuestionKey: previousQuestionKey,
                    keyFifths: keyFifths
                ) else { continue }
                return built

            case .progression:
                continue
            }
        }

        return buildQuestion(options: TrainingQuestionBuilderOptions(
            training: training,
            ignoreNotationInstrument: options.ignoreNotationInstrument,
            lessonRoots: options.lessonRoots,
            lessonOrder: options.lessonOrder,
            lessonItems: options.lessonItems,
            lessonItemIndex: options.lessonItemIndex,
            previousQuestionKey: nil
        ))
    }

    /// ステージ内で出題しうる全 MIDI。音域フィットを問題ごとに動かさないために使う。
    static func collectStageMidis(training: TrainingRow) -> [Int] {
        let mergedConfig = training.config
        let roots = mergedConfig.roots ?? ["C"]
        let effectiveClef = resolveEffectiveClef(clefMode: training.clefMode, configClef: mergedConfig.clef)
        let singleClef = effectiveClef == "bass" ? "bass" : "treble"
        let staffBottom = staffBottomMidi[singleClef] ?? 64
        var midis: [Int] = []

        switch training.kind {
        case .noteReading:
            let includeAccidentals = mergedConfig.includeAccidentals == true
            let forcedClef = mergedConfig.clef == "bass" || effectiveClef == "bass" ? "bass" : "treble"
            let spellings = includeAccidentals
                ? noteReadingWithAccidentals[forcedClef]
                : noteReadingNaturals[forcedClef]
            if let spellings {
                for spelling in spellings {
                    if let midi = TrainingMusicTheory.parseVoicingMidi(spelling) {
                        midis.append(midi)
                    }
                }
            }

        case .interval:
            let interval = mergedConfig.interval ?? "2m"
            let directionUp = (mergedConfig.direction ?? "up") == "up"
            guard let spec = TrainingMusicTheory.parseInterval(directionUp ? interval : "-\(interval)") else {
                return midis
            }
            for spelling in intervalBaseSpellings {
                guard let raw = TrainingMusicTheory.parseSpelled("\(spelling)4") else { continue }
                guard let base = TrainingMusicTheory.placeLowestInOctaveAbove([raw], minMidi: staffBottom).first else {
                    continue
                }
                let target = TrainingMusicTheory.transpose(base, by: spec)
                guard TrainingMusicTheory.isSimpleSpelling(target) else { continue }
                if target.midi < staffBottom {
                    midis.append(TrainingMusicTheory.shiftOctave(base, by: 1).midi)
                    midis.append(TrainingMusicTheory.shiftOctave(target, by: 1).midi)
                } else {
                    midis.append(base.midi)
                    midis.append(target.midi)
                }
            }

        case .scale:
            let scaleType = mergedConfig.scale ?? "major"
            guard let intervals = TrainingMusicTheory.scaleTemplates[scaleType] else { return midis }
            for root in roots {
                let notes = TrainingMusicTheory.placeLowestInOctaveAbove(
                    TrainingMusicTheory.spelledFromIntervals(root: root, octave: 4, intervals: intervals),
                    minMidi: staffBottom
                )
                midis.append(contentsOf: notes.map(\.midi))
            }

        case .chord, .voicing:
            let defaultStaff = singleClef == "bass" ? 2 : 1
            for root in roots {
                guard let built = buildChordVoicingQuestion(
                    training: training,
                    config: mergedConfig,
                    root: root,
                    defaultStaff: defaultStaff,
                    staffBottom: staffBottom,
                    previousQuestionKey: nil,
                    keyFifths: 0
                ) else { continue }
                midis.append(contentsOf: built.notes.map(\.midi))
            }

        case .progression:
            break
        }

        return midis
    }

    private static func buildChordVoicingQuestion(
        training: TrainingRow,
        config: TrainingConfig,
        root: String,
        defaultStaff: Int,
        staffBottom: Int,
        previousQuestionKey: String?,
        keyFifths: Int
    ) -> TrainingQuestion? {
        let notes: [SpelledNote]
        let staves: [Int]

        if let voicingNotes = config.voicingNotes, !voicingNotes.isEmpty {
            let parsed = voicingNotes.compactMap { TrainingMusicTheory.parseSpelled($0) }
            guard parsed.count == voicingNotes.count else { return nil }
            let transposed: [SpelledNote]
            if let interval = TrainingMusicTheory.ascendingInterval(fromRoot: config.referenceRoot ?? "C", toRoot: root),
               interval.semitones != 0 || interval.steps != 0 {
                transposed = parsed.map { TrainingMusicTheory.transpose($0, by: interval) }
            } else {
                transposed = parsed
            }
            let minMidi = config.minLowestNote.flatMap { TrainingMusicTheory.parseVoicingMidi($0) } ?? staffBottom
            notes = TrainingMusicTheory.placeLowestInOctaveAbove(transposed, minMidi: minMidi)
            staves = toStaves(config.staves, count: notes.count, fallback: defaultStaff)
        } else if let intervals = config.intervals, !intervals.isEmpty {
            let raw = TrainingMusicTheory.spelledFromIntervals(root: root, octave: 3, intervals: intervals)
            let minMidi = config.minLowestNote.flatMap { TrainingMusicTheory.parseVoicingMidi($0) } ?? staffBottom
            notes = TrainingMusicTheory.placeLowestInOctaveAbove(raw, minMidi: minMidi)
            staves = toStaves(config.staves, count: notes.count, fallback: defaultStaff)
        } else if let quality = config.quality, let intervals = TrainingMusicTheory.chordTemplates[quality] {
            notes = TrainingMusicTheory.placeLowestInOctaveAbove(
                TrainingMusicTheory.spelledFromIntervals(root: root, octave: 4, intervals: intervals),
                minMidi: staffBottom
            )
            staves = toStaves(nil, count: notes.count, fallback: defaultStaff)
        } else {
            return nil
        }
        guard !notes.isEmpty, let lowestMidi = notes.map(\.midi).min() else { return nil }

        let noteNames = notes.map(\.name)
        let questionKey = "chord:\(root):\(noteNames.joined(separator: "|"))"
        if questionKey == previousQuestionKey { return nil }
        let suffix: String
        if training.kind == .chord, let quality = config.quality {
            suffix = TrainingMusicTheory.chordSymbolSuffix[quality] ?? quality
        } else {
            suffix = training.titleEn
        }
        return makeQuestion(
            questionKey: questionKey,
            promptLabel: "\(root)\(suffix)",
            noteNames: noteNames,
            staves: staves,
            targets: noteNames.map { _ in true },
            layout: .stacked,
            ordered: false,
            keyFifths: keyFifths,
            rootMidi: TrainingMusicTheory.rootMidiBelow(root: root, lowestMidi: lowestMidi)
        )
    }

    private static func mergeConfig(base: TrainingConfig, override: TrainingConfig) -> TrainingConfig {
        TrainingConfig(
            roots: override.roots ?? base.roots,
            quality: override.quality ?? base.quality,
            scale: override.scale ?? base.scale,
            interval: override.interval ?? base.interval,
            direction: override.direction ?? base.direction,
            clef: override.clef ?? base.clef,
            includeAccidentals: override.includeAccidentals ?? base.includeAccidentals,
            intervals: override.intervals ?? base.intervals,
            staves: override.staves ?? base.staves,
            voicingNotes: override.voicingNotes ?? base.voicingNotes,
            referenceRoot: override.referenceRoot ?? base.referenceRoot,
            minLowestNote: override.minLowestNote ?? base.minLowestNote
        )
    }

    private static func resolveEffectiveClef(clefMode: TrainingClefMode, configClef: String?) -> String {
        switch clefMode {
        case .bassConcert: return "bass"
        case .grandConcert: return "grand"
        case .instrument:
            if configClef == "bass" { return "bass" }
            return "treble"
        }
    }

    private static func toStaves(_ staves: [Int]?, count: Int, fallback: Int) -> [Int] {
        (0..<count).map { index in
            guard let staves, index < staves.count else { return fallback }
            return staves[index] == 2 ? 2 : (staves[index] == 1 ? 1 : fallback)
        }
    }

    private static func makeQuestion(
        questionKey: String,
        promptLabel: String,
        noteNames: [String],
        staves: [Int],
        targets: [Bool],
        layout: TrainingQuestionLayout,
        ordered: Bool,
        keyFifths: Int,
        rootMidi: Int?
    ) -> TrainingQuestion {
        let notes: [TrainingQuestionNote] = noteNames.enumerated().compactMap { index, name in
            guard let midi = TrainingMusicTheory.parseVoicingMidi(name) else { return nil }
            return TrainingQuestionNote(
                noteName: name,
                midi: midi,
                pitchClass: TrainingMusicTheory.normalizePitchClass(midi),
                staff: index < staves.count ? staves[index] : 1,
                isTarget: index < targets.count ? targets[index] : true
            )
        }
        return TrainingQuestion(
            questionKey: questionKey,
            promptLabel: promptLabel,
            notes: notes,
            layout: layout,
            ordered: ordered,
            keyFifths: keyFifths,
            rootMidi: rootMidi
        )
    }
}
